File: cloakbrowser/human/actionability.py
"""Playwright-style actionability checks for the humanize layer (sync).

Checks: attached, visible, stable, enabled, editable, receives pointer events.
Retry loop with backoff matching Playwright internals: [100, 250, 500, 1000]ms.
"""

from __future__ import annotations

import time
from typing import Any, FrozenSet, Optional

from .stealth_dom import (
    build_actionable_js, build_box_js, build_validate_js, eval_parsed,
    EVALUATION_FAILED, NOT_FOUND, OK, STALE, UNSUPPORTED,
    StealthEvaluationError, StealthWorldUnavailableError,
    UnsupportedHumanizeSelectorError,
)

# ---------------------------------------------------------------------------
# Error hierarchy — all subclass RuntimeError for backward compat
# ---------------------------------------------------------------------------

class ActionabilityError(RuntimeError):
    """Base for all actionability failures."""

    def __init__(self, selector: str, check: str, message: str):
        self.selector = selector
        self.check = check
        super().__init__(f"Element {selector!r} failed {check} check: {message}")


class ElementNotAttachedError(ActionabilityError):
    def __init__(self, selector: str):
        super().__init__(selector, "attached", "element not found in DOM")


class ElementNotVisibleError(ActionabilityError):
    def __init__(self, selector: str):
        super().__init__(selector, "visible", "element is not visible")


class ElementNotStableError(ActionabilityError):
    def __init__(self, selector: str):
        super().__init__(selector, "stable", "element position is still changing")


class ElementNotEnabledError(ActionabilityError):
    def __init__(self, selector: str):
        super().__init__(selector, "enabled", "element is disabled")


class ElementNotEditableError(ActionabilityError):
    def __init__(self, selector: str):
        super().__init__(selector, "editable", "element is not editable")


class ElementNotReceivingEventsError(ActionabilityError):
    def __init__(self, selector: str, covering_tag: str = "unknown"):
        super().__init__(
            selector,
            "pointer_events",
            f"element is covered by <{covering_tag}>",
        )


class ElementTargetChangedError(ActionabilityError):
    def __init__(self, selector: str):
        super().__init__(
            selector,
            "target_identity",
            "resolved element changed before pointer dispatch",
        )


# ---------------------------------------------------------------------------
# Check-set constants
# ---------------------------------------------------------------------------

CHECKS_CLICK: FrozenSet[str] = frozenset({"attached", "visible", "enabled", "pointer_events"})
CHECKS_HOVER: FrozenSet[str] = frozenset({"attached", "visible", "pointer_events"})
CHECKS_INPUT: FrozenSet[str] = frozenset({"attached", "visible", "enabled", "editable", "pointer_events"})
CHECKS_FOCUS: FrozenSet[str] = frozenset({"attached", "visible", "enabled"})
CHECKS_CHECK: FrozenSet[str] = frozenset({"attached", "visible", "enabled", "pointer_events"})

_BACKOFF_MS = [100, 250, 500, 1000]


def _backoff_sleep(attempt: int) -> None:
    idx = min(attempt, len(_BACKOFF_MS) - 1)
    time.sleep(_BACKOFF_MS[idx] / 1000.0)


# ---------------------------------------------------------------------------
# Pre-scroll actionability: attached, visible, enabled, editable
# ---------------------------------------------------------------------------

def _stealth_actionable(page: Any, selector: str, checks: FrozenSet[str]) -> None:
    """Run actionability checks exclusively through the isolated world."""
    world = getattr(page, "_stealth_world", None)
    if world is None:
        raise StealthWorldUnavailableError()
    status, data = eval_parsed(world, build_actionable_js(selector))
    if status == UNSUPPORTED:
        raise UnsupportedHumanizeSelectorError(selector)
    if status == EVALUATION_FAILED:
        raise StealthEvaluationError(selector)
    if status == NOT_FOUND:
        # Every check-set includes 'attached'; not present yet -> raise so the
        # retry loop backs off and re-reads in-world (mirrors wait_for(attached)).
        raise ElementNotAttachedError(selector)
    if "visible" in checks and not data.get("visible"):
        raise ElementNotVisibleError(selector)
    if "enabled" in checks and not data.get("enabled"):
        raise ElementNotEnabledError(selector)
    if "editable" in checks and not data.get("editable"):
        raise ElementNotEditableError(selector)


def _read_box(page: Any, selector: str, remaining_ms: float) -> Optional[dict]:
    """Read geometry exclusively through the isolated world."""
    world = getattr(page, "_stealth_world", None)
    if world is None:
        raise StealthWorldUnavailableError()
    status, data = eval_parsed(world, build_box_js(selector))
    if status == OK:
        return data["box"]
    if status == NOT_FOUND:
        return None
    if status == UNSUPPORTED:
        raise UnsupportedHumanizeSelectorError(selector)
    raise StealthEvaluationError(selector)


def ensure_actionable(
    page: Any,
    selector: str,
    checks: FrozenSet[str],
    timeout: float = 30000,
    force: bool = False,
) -> None:
    """Wait for element to pass actionability checks (pre-scroll).

    Retries with backoff until *timeout* ms elapsed.
    Raises a specific ``ActionabilityError`` subclass on failure.
    If *force* is True, returns immediately.
    """
    if force:
        return

    deadline = time.monotonic() + timeout / 1000.0
    attempt = 0
    last_error: Optional[ActionabilityError] = None

    while True:
        remaining_ms = max(0, (deadline - time.monotonic()) * 1000)
        if remaining_ms <= 0:
            if last_error is not None:
                raise last_error
            raise ActionabilityError(selector, "timeout", "timeout expired before first check")

        try:
            _stealth_actionable(page, selector, checks)
            return

        except (ActionabilityError, StealthEvaluationError) as e:
            last_error = e
            if time.monotonic() >= deadline:
                raise last_error
            _backoff_sleep(attempt)
            attempt += 1


# ---------------------------------------------------------------------------
# Post-scroll stability check
# ---------------------------------------------------------------------------

def _boxes_differ(a: dict, b: dict) -> bool:
    return (
        abs(a["x"] - b["x"]) > 1
        or abs(a["y"] - b["y"]) > 1
        or abs(a["width"] - b["width"]) > 1
        or abs(a["height"] - b["height"]) > 1
    )


def ensure_stable(
    page: Any,
    selector: str,
    timeout: float = 5000,
) -> None:
    """Wait for element position to stabilize (two samples 100ms apart).

    Only call after scroll — skip if element was already in viewport.
    """
    deadline = time.monotonic() + timeout / 1000.0
    attempt = 0

    while True:
        remaining_ms = max(0, (deadline - time.monotonic()) * 1000)
        if remaining_ms <= 0:
            raise ElementNotStableError(selector)

        try:
            box1 = _read_box(page, selector, remaining_ms)
            if box1 is None:
                raise ElementNotAttachedError(selector)

            time.sleep(0.1)

            box2 = _read_box(page, selector, remaining_ms)
            if box2 is None:
                raise ElementNotAttachedError(selector)
        except StealthEvaluationError:
            if time.monotonic() >= deadline:
                raise
            _backoff_sleep(attempt)
            attempt += 1
            continue

        if not _boxes_differ(box1, box2):
            return

        if time.monotonic() >= deadline:
            raise ElementNotStableError(selector)

        _backoff_sleep(attempt)
        attempt += 1


# ---------------------------------------------------------------------------
# Pointer-events check (post-scroll, at actual click coordinates)
# ---------------------------------------------------------------------------

# Legacy ElementHandle paths cannot use selector-based isolated-world identity.
_POINTER_EVENTS_HANDLE_JS = """(expected, data) => {
    const rect = expected.getBoundingClientRect();
    const frameOffsetX = data.box ? data.box.x - rect.x : 0;
    const frameOffsetY = data.box ? data.box.y - rect.y : 0;
    const target = document.elementFromPoint(data.x - frameOffsetX, data.y - frameOffsetY);
    if (!target) return { hit: false, reason: 'no_element_at_point', covering: 'none' };
    let node = target;
    while (node) { if (node === expected) return { hit: true }; node = node.parentNode; }
    if (expected.contains(target)) return { hit: true };
    return { hit: false, reason: 'covered', covering: target.tagName || 'unknown' };
}"""


def check_pointer_events(
    page: Any,
    selector: str,
    target_id: int,
    gen: int,
    x: float,
    y: float,
    stealth: Any = None,
    timeout: float = 5000,
) -> None:
    """Revalidate that the click point still hits the same resolved element.

    Callers skip this entirely when ``force=True`` (matching Playwright, where
    ``force`` bypasses all actionability checks).
    """
    deadline = time.monotonic() + timeout / 1000.0
    attempt = 0
    last_miss: Optional[str] = None
    world = stealth if stealth is not None else getattr(page, "_stealth_world", None)
    if world is None:
        raise StealthWorldUnavailableError()
    if not isinstance(target_id, int) or not isinstance(gen, int):
        raise StealthEvaluationError(selector)

    while True:
        status, data = eval_parsed(
            world, build_validate_js(selector, target_id, gen, x, y)
        )
        if status == UNSUPPORTED:
            raise UnsupportedHumanizeSelectorError(selector)
        if status == STALE:
            raise ElementTargetChangedError(selector)
        if status == NOT_FOUND:
            raise ElementNotAttachedError(selector)
        if status == EVALUATION_FAILED:
            if time.monotonic() >= deadline:
                raise StealthEvaluationError(selector)
        elif data.get("hit", False):
            return
        else:
            last_miss = data.get("covering", "unknown")
            if time.monotonic() >= deadline:
                raise ElementNotReceivingEventsError(selector, last_miss)

        _backoff_sleep(attempt)
        attempt += 1


# ---------------------------------------------------------------------------
# ElementHandle variant
# ---------------------------------------------------------------------------

def ensure_actionable_handle(
    page: Any,
    el: Any,
    checks: FrozenSet[str],
    timeout: float = 30000,
    force: bool = False,
) -> None:
    """Actionability checks for ElementHandle (no selector needed).

    Uses Playwright's wait_for_element_state where available.
    """
    if force:
        return

    deadline = time.monotonic() + timeout / 1000.0
    attempt = 0
    last_error: Optional[ActionabilityError] = None
    label = "<ElementHandle>"

    while True:
        remaining_ms = max(0, (deadline - time.monotonic()) * 1000)
        if remaining_ms <= 0:
            if last_error is not None:
                raise last_error
            raise ActionabilityError(label, "timeout", "timeout expired before first check")

        try:
            if "visible" in checks:
                try:
                    el.wait_for_element_state("visible", timeout=max(1, min(remaining_ms, 2000)))
                except Exception:
                    raise ElementNotVisibleError(label)

            if "enabled" in checks:
                try:
                    el.wait_for_element_state("enabled", timeout=max(1, min(remaining_ms, 2000)))
                except Exception:
                    raise ElementNotEnabledError(label)

            if "editable" in checks:
                try:
                    el.wait_for_element_state("editable", timeout=max(1, min(remaining_ms, 2000)))
                except Exception:
                    raise ElementNotEditableError(label)

            return

        except ActionabilityError as e:
            last_error = e
            if time.monotonic() >= deadline:
                raise last_error
            _backoff_sleep(attempt)
            attempt += 1


def check_pointer_events_handle(
    page: Any,
    el: Any,
    x: float,
    y: float,
    timeout: float = 5000,
) -> None:
    """Pointer-events check for ElementHandle."""
    deadline = time.monotonic() + timeout / 1000.0
    attempt = 0

    while True:
        try:
            box = el.bounding_box()
            result = el.evaluate(_POINTER_EVENTS_HANDLE_JS, {"x": x, "y": y, "box": box})
        except Exception:
            result = None

        # Proceed if the check confirms a hit, or if it could not be determined
        # (None) — failing closed would block legitimate clicks.
        if result is None or result.get("hit", False):
            return

        covering = (result or {}).get("covering", "unknown")

        if time.monotonic() >= deadline:
            raise ElementNotReceivingEventsError("<ElementHandle>", covering)

        _backoff_sleep(attempt)
        attempt += 1

File: js/src/human/actionability.ts
/**
 * Playwright-style actionability checks for the humanize layer.
 *
 * Selector-based reads stay inside the CDP isolated world. ElementHandle paths
 * remain separate because they already identify an exact Playwright handle.
 */

import type { Page, Frame, ElementHandle } from 'playwright-core';
import {
  buildActionableJs, buildBoxJs, buildValidateJs, evalParsed, getWorld,
  OK, NOT_FOUND, UNSUPPORTED, STALE, EVALUATION_FAILED,
  StealthEvaluationError, StealthWorldUnavailableError,
  UnsupportedHumanizeSelectorError, type StealthWorld,
} from './stealthDom.js';

// ---------------------------------------------------------------------------
// Error hierarchy
// ---------------------------------------------------------------------------

export class ActionabilityError extends Error {
  selector: string;
  check: string;

  constructor(selector: string, check: string, message: string) {
    super(`Element ${JSON.stringify(selector)} failed ${check} check: ${message}`);
    this.name = 'ActionabilityError';
    this.selector = selector;
    this.check = check;
  }
}

export class ElementNotAttachedError extends ActionabilityError {
  constructor(selector: string) {
    super(selector, 'attached', 'element not found in DOM');
    this.name = 'ElementNotAttachedError';
  }
}

export class ElementNotVisibleError extends ActionabilityError {
  constructor(selector: string) {
    super(selector, 'visible', 'element is not visible');
    this.name = 'ElementNotVisibleError';
  }
}

export class ElementNotStableError extends ActionabilityError {
  constructor(selector: string) {
    super(selector, 'stable', 'element position is still changing');
    this.name = 'ElementNotStableError';
  }
}

export class ElementNotEnabledError extends ActionabilityError {
  constructor(selector: string) {
    super(selector, 'enabled', 'element is disabled');
    this.name = 'ElementNotEnabledError';
  }
}

export class ElementNotEditableError extends ActionabilityError {
  constructor(selector: string) {
    super(selector, 'editable', 'element is not editable');
    this.name = 'ElementNotEditableError';
  }
}

export class ElementNotReceivingEventsError extends ActionabilityError {
  coveringTag: string;
  constructor(selector: string, coveringTag: string = 'unknown') {
    super(selector, 'pointer_events', `element is covered by <${coveringTag}>`);
    this.name = 'ElementNotReceivingEventsError';
    this.coveringTag = coveringTag;
  }
}

export class ElementTargetChangedError extends ActionabilityError {
  constructor(selector: string) {
    super(selector, 'target_identity', 'selector resolved to a different element before input dispatch');
    this.name = 'ElementTargetChangedError';
  }
}

// ---------------------------------------------------------------------------
// Check-set constants
// ---------------------------------------------------------------------------

export type CheckName = 'attached' | 'visible' | 'enabled' | 'editable' | 'pointer_events';

export const CHECKS_CLICK: ReadonlySet<CheckName> = new Set(['attached', 'visible', 'enabled', 'pointer_events']);
export const CHECKS_HOVER: ReadonlySet<CheckName> = new Set(['attached', 'visible', 'pointer_events']);
export const CHECKS_INPUT: ReadonlySet<CheckName> = new Set(['attached', 'visible', 'enabled', 'editable', 'pointer_events']);
export const CHECKS_FOCUS: ReadonlySet<CheckName> = new Set(['attached', 'visible', 'enabled']);
export const CHECKS_CHECK: ReadonlySet<CheckName> = new Set(['attached', 'visible', 'enabled', 'pointer_events']);

const BACKOFF_MS = [100, 250, 500, 1000];

function backoffSleep(attempt: number): Promise<void> {
  const idx = Math.min(attempt, BACKOFF_MS.length - 1);
  return new Promise(resolve => setTimeout(resolve, BACKOFF_MS[idx]));
}

// ---------------------------------------------------------------------------
// Pre-scroll actionability
// ---------------------------------------------------------------------------

async function stealthActionable(
  pageOrFrame: Page | Frame,
  selector: string,
  checks: ReadonlySet<CheckName>,
): Promise<void> {
  const world = getWorld(pageOrFrame);
  if (!world) throw new StealthWorldUnavailableError();

  const { status, data } = await evalParsed(world, buildActionableJs(selector));
  if (status === UNSUPPORTED) throw new UnsupportedHumanizeSelectorError(selector);
  if (status === EVALUATION_FAILED) throw new StealthEvaluationError(selector);
  if (status === NOT_FOUND) throw new ElementNotAttachedError(selector);
  if (status !== OK || !data) throw new StealthEvaluationError(selector);
  if (checks.has('visible') && !data.visible) throw new ElementNotVisibleError(selector);
  if (checks.has('enabled') && !data.enabled) throw new ElementNotEnabledError(selector);
  if (checks.has('editable') && !data.editable) throw new ElementNotEditableError(selector);
}

async function readBox(
  pageOrFrame: Page | Frame,
  selector: string,
): Promise<{ x: number; y: number; width: number; height: number } | null> {
  const world = getWorld(pageOrFrame);
  if (!world) throw new StealthWorldUnavailableError();

  const { status, data } = await evalParsed(world, buildBoxJs(selector));
  if (status === OK && data?.box) return data.box;
  if (status === NOT_FOUND) return null;
  if (status === UNSUPPORTED) throw new UnsupportedHumanizeSelectorError(selector);
  throw new StealthEvaluationError(selector);
}

export async function ensureActionable(
  pageOrFrame: Page | Frame,
  selector: string,
  checks: ReadonlySet<CheckName>,
  timeout: number = 30000,
  force: boolean = false,
): Promise<void> {
  if (force) return;

  const deadline = Date.now() + timeout;
  let attempt = 0;
  let lastError: Error | null = null;

  while (true) {
    const remainingMs = Math.max(0, deadline - Date.now());
    if (remainingMs <= 0) {
      if (lastError) throw lastError;
      throw new ActionabilityError(selector, 'timeout', 'timeout expired before first check');
    }

    try {
      await stealthActionable(pageOrFrame, selector, checks);
      return;
    } catch (error) {
      if (error instanceof ActionabilityError || error instanceof StealthEvaluationError) {
        lastError = error;
        if (Date.now() >= deadline) throw lastError;
        await backoffSleep(attempt++);
      } else {
        throw error;
      }
    }
  }
}

// ---------------------------------------------------------------------------
// Post-scroll stability check
// ---------------------------------------------------------------------------

function boxesDiffer(
  a: { x: number; y: number; width: number; height: number },
  b: { x: number; y: number; width: number; height: number },
): boolean {
  return (
    Math.abs(a.x - b.x) > 1 ||
    Math.abs(a.y - b.y) > 1 ||
    Math.abs(a.width - b.width) > 1 ||
    Math.abs(a.height - b.height) > 1
  );
}

export async function ensureStable(
  pageOrFrame: Page | Frame,
  selector: string,
  timeout: number = 5000,
): Promise<void> {
  const deadline = Date.now() + timeout;
  let attempt = 0;

  while (true) {
    const remainingMs = Math.max(0, deadline - Date.now());
    if (remainingMs <= 0) throw new ElementNotStableError(selector);

    try {
      const box1 = await readBox(pageOrFrame, selector);
      if (!box1) throw new ElementNotAttachedError(selector);

      await new Promise(resolve => setTimeout(resolve, 100));

      const box2 = await readBox(pageOrFrame, selector);
      if (!box2) throw new ElementNotAttachedError(selector);
      if (!boxesDiffer(box1, box2)) return;
    } catch (error) {
      if (error instanceof StealthEvaluationError) {
        if (Date.now() >= deadline) throw error;
        await backoffSleep(attempt++);
        continue;
      }
      throw error;
    }

    if (Date.now() >= deadline) throw new ElementNotStableError(selector);
    await backoffSleep(attempt++);
  }
}

// ---------------------------------------------------------------------------
// Pointer-events and exact-target check
// ---------------------------------------------------------------------------

const POINTER_EVENTS_HANDLE_JS = `(expected, data) => {
  const rect = expected.getBoundingClientRect();
  const frameOffsetX = data.box ? data.box.x - rect.x : 0;
  const frameOffsetY = data.box ? data.box.y - rect.y : 0;
  const target = document.elementFromPoint(data.x - frameOffsetX, data.y - frameOffsetY);
  if (!target) return { hit: false, reason: 'no_element_at_point', covering: 'none' };
  let node = target;
  while (node) { if (node === expected) return { hit: true }; node = node.parentNode; }
  if (expected.contains(target)) return { hit: true };
  return { hit: false, reason: 'covered', covering: target.tagName || 'unknown' };
}`;

export async function checkPointerEvents(
  pageOrFrame: Page | Frame,
  selector: string,
  targetId: number,
  gen: number,
  x: number,
  y: number,
  stealth?: StealthWorld | null,
  timeout: number = 5000,
): Promise<void> {
  const deadline = Date.now() + timeout;
  let attempt = 0;
  let lastMiss: string | null = null;
  const world = stealth ?? getWorld(pageOrFrame);

  if (!world) throw new StealthWorldUnavailableError();
  if (!Number.isInteger(targetId) || !Number.isInteger(gen)) throw new StealthEvaluationError(selector);

  while (true) {
    const { status, data } = await evalParsed(
      world,
      buildValidateJs(selector, targetId, gen, x, y),
    );

    if (status === UNSUPPORTED) throw new UnsupportedHumanizeSelectorError(selector);
    if (status === STALE) throw new ElementTargetChangedError(selector);
    if (status === NOT_FOUND) throw new ElementNotAttachedError(selector);
    if (status === EVALUATION_FAILED) {
      if (Date.now() >= deadline) throw new StealthEvaluationError(selector);
    } else if (status === OK && data && data.hit) {
      return;
    } else if (status === OK && data) {
      lastMiss = data.covering ?? 'unknown';
      if (Date.now() >= deadline) {
        throw new ElementNotReceivingEventsError(selector, lastMiss ?? 'unknown');
      }
    } else {
      throw new StealthEvaluationError(selector);
    }

    await backoffSleep(attempt++);
  }
}

// ---------------------------------------------------------------------------
// ElementHandle variant (legacy handle-scoped path)
// ---------------------------------------------------------------------------

export async function ensureActionableHandle(
  el: ElementHandle,
  checks: ReadonlySet<CheckName>,
  timeout: number = 30000,
  force: boolean = false,
): Promise<void> {
  if (force) return;

  const deadline = Date.now() + timeout;
  let attempt = 0;
  let lastError: ActionabilityError | null = null;
  const label = '<ElementHandle>';

  while (true) {
    const remainingMs = Math.max(0, deadline - Date.now());
    if (remainingMs <= 0) {
      if (lastError) throw lastError;
      throw new ActionabilityError(label, 'timeout', 'timeout expired before first check');
    }

    try {
      if (checks.has('visible')) {
        try {
          await el.waitForElementState('visible', { timeout: Math.max(1, Math.min(remainingMs, 2000)) });
        } catch {
          throw new ElementNotVisibleError(label);
        }
      }
      if (checks.has('enabled')) {
        try {
          await el.waitForElementState('enabled', { timeout: Math.max(1, Math.min(remainingMs, 2000)) });
        } catch {
          throw new ElementNotEnabledError(label);
        }
      }
      if (checks.has('editable')) {
        try {
          await el.waitForElementState('editable', { timeout: Math.max(1, Math.min(remainingMs, 2000)) });
        } catch {
          throw new ElementNotEditableError(label);
        }
      }
      return;
    } catch (error) {
      if (error instanceof ActionabilityError) {
        lastError = error;
        if (Date.now() >= deadline) throw lastError;
        await backoffSleep(attempt++);
      } else {
        throw error;
      }
    }
  }
}

export async function checkPointerEventsHandle(
  el: ElementHandle,
  x: number,
  y: number,
  timeout: number = 5000,
): Promise<void> {
  const deadline = Date.now() + timeout;
  let attempt = 0;

  while (true) {
    let result: any;
    try {
      const box = await el.boundingBox();
      result = await el.evaluate(POINTER_EVENTS_HANDLE_JS, { x, y, box });
    } catch {
      result = null;
    }

    if (!result || result.hit) return;

    const covering = result?.covering ?? 'unknown';
    if (Date.now() >= deadline) {
      throw new ElementNotReceivingEventsError('<ElementHandle>', covering);
    }
    await backoffSleep(attempt++);
  }
}
